src/lib/corchuelo.rs

```diff
@@ -84,233 +120,356 @@ impl<TokId: PrimInt + Unsigned> Recoverer<TokId> for Corchuelo
+                        continue;
                     }
-                    todo.push_back((new_la_idx, n_pstack, n_repairs, sc));
+                }
+
+                let next_lexeme = self.parser.next_lexeme(n.la_idx);
+                let new_lexeme = Lexeme::new(TokId::from(u32::from(t_idx)).unwrap(),
+                                             next_lexeme.start(), 0);
+                let (new_la_idx, n_pstack) =
+                    self.parser.lr_cactus(Some(new_lexeme), la_idx, la_idx + 1,
+                                          n.pstack.clone(), &mut None);
+                if new_la_idx > la_idx {
+                    let n_repairs = n.repairs.child(Repair::InsertTerm(t_idx));
+                    let nn = PathFNode{
+                        pstack: n_pstack,
+                        la_idx: n.la_idx,
+                        repairs: n_repairs,
+                        cf: n.cf.checked_add((self.parser.term_cost)(t_idx) as u32).unwrap()};
+                    nbrs.push((nn.cf, nn));
                 }
             }
         }
+    }
 
-        let repairs = finished.iter()
-                              .map(|x| { let mut v = x.vals().cloned().collect::<Vec<Repair>>();
-                                         v.reverse();
-                                         v
-                               })
-                              .collect::<Vec<Vec<Repair>>>();
+    fn delete(&self,
+           n: &PathFNode,
+           nbrs: &mut Vec<(u32, PathFNode)>)
+    {
+        if n.la_idx == self.parser.lexemes.len() {
+            return;
+        }
+
+        let n_repairs = n.repairs.child(Repair::Delete);
+        let la_tidx = self.parser.next_tidx(n.la_idx);
+        let cost = (self.parser.term_cost)(la_tidx);
+        let nn = PathFNode{pstack: n.pstack.clone(),
+                           la_idx: n.la_idx + 1,
+                           repairs: n_repairs,
+                           cf: n.cf.checked_add(cost as u32).unwrap()};
+        nbrs.push((nn.cf, nn));
+    }
 
-        if repairs.is_empty() {
-            return (in_la_idx, Vec::new());
+    fn shift(&self,
+             n: &PathFNode,
+             nbrs: &mut Vec<(u32, PathFNode)>)
+    {
+        // Forward move rule (ER3)
+        //
+        // Note the rule in Corchuelo et al. is confusing and, I think, wrong. It reads:
+        //   (S, I) \rightarrow_{LR*} (S', I')
+        //   \wedge (j = N \vee 0 < j < N \wedge f(q_r, t_{j + 1} \in {accept, error})
+        // First I think the bracketing would be clearer if written as:
+        //   j = N \vee (0 < j < N \wedge f(q_r, t_{j + 1} \in {accept, error})
+        // And I think the condition should be:
+        //   j = N \vee (0 <= j < N \wedge f(q_r, t_{j + 1} \in {accept, error})
+        // because there's no reason that any symbols need to be shifted in order for an accept
+        // (or, indeed an error) state to be reached.
+        //
+        // So the full rule should I think be:
```

Comment:
Better? "So the full rule I think should be"

Reply:
Probably best is "So the full should, I think, be:" (I missed the commas!)?

Reply:
Commas is also fine.